Uploaded Artists: fix fetching and displaying uploaded artists

mattcarter11 opened the pull request:
Before, you could not fetch and display the privately owned (uploaded) artists

<p><strong>Uploaded artist</strong></p>
<img src="https://github.com/user-attachments/assets/ab825ba6-5ff9-4c2c-be0a-76a167ef1ae7" width="200"/>
<img src="https://github.com/user-attachments/assets/52a311b7-b62b-482f-b061-40e5ffdb81b7" width="200"/>

<p><strong>YouTube artist</strong></p>
<img src="https://github.com/user-attachments/assets/56570915-927d-4edf-8016-47dcee194f85" width="200"/>
<img src="https://github.com/user-attachments/assets/fc518902-6771-4a2a-a7d2-cab6552f32fe" width="200"/>

Also fixed:
- not all chips syncing when being selected
- don't sync if no connection. Finish sync if no error while fetching ytm data
- when syncing playlist, exclude liked playlist by id, not by order
-  change text sort by date to sort by liked when songs liked filter is active

mattcarter11 committed:
Uploaded Artists: fix fetching and displaying uploaded artists

mattcarter11 commented:
@mikooomich a fix from a missing piece from the big PR

mattcarter11 committed:
albums: add local screen for artists (and fix some artists not showing up)

mattcarter11 committed:
albums: add local artist album view

mattcarter11 committed:
chips: some chips didn't sync when clicking

mikooomich commented:
Do you know why it crashes when trying open pages for local artists?





```

Process: com.dd3boh.outertune.debug, PID: 26517
java.lang.NullPointerException: Parameter specified as non-null is null: method androidx.room.RoomSQLiteQuery.bindString, parameter value
	at androidx.room.RoomSQLiteQuery.bindString(Unknown Source:3)
	at com.dd3boh.outertune.db.DatabaseDao_Impl.__fetchRelationshipartistAscomDd3bohOutertuneDbEntitiesArtistEntity_1(DatabaseDao_Impl.java:12524)
	at com.dd3boh.outertune.db.DatabaseDao_Impl.-$$Nest$m__fetchRelationshipartistAscomDd3bohOutertuneDbEntitiesArtistEntity_1(Unknown Source:0)
	at com.dd3boh.outertune.db.DatabaseDao_Impl$112.call(DatabaseDao_Impl.java:9829)
	at com.dd3boh.outertune.db.DatabaseDao_Impl$112.call(DatabaseDao_Impl.java:9802)
	at androidx.room.CoroutinesRoom$Companion$createFlow$1$1$1.invokeSuspend(CoroutinesRoom.kt:129)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:104)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:644)
	at java.lang.Thread.run(Thread.java:1012)
	Suppressed: kotlinx.coroutines.internal.DiagnosticCoroutineContextException: 
```


For example, local artists look like this where the are some null fields, and isLocal = 1
![image](https://github.com/user-attachments/assets/94722042-61df-497b-9025-1ae925f1c1d9)



The page before this pr
![image](https://github.com/user-attachments/assets/d17bab1c-faaf-4dfe-9885-479690efe555)

mattcarter11 committed:
sync: don't sync if no connection. Finish sync if no error while fetching ytm data

mattcarter11 commented:
@mikooomich i'll take a look.

Also here I've added two sync fixes:
- not all chips syncing when being selected
- don't sync if no connection. Finish sync if no error while fetching ytm data

mattcarter11 commented:
@mikooomich I think I've found the problem and fixed it!

mattcarter11 committed:
songs: change text when liked filter: sort by date -> sort by liked

mattcarter11 committed:
sync playlists: exclude liked by id, not order

mattcarter11 commented:
@mikooomich also, after adding the corutines to sync songs faster. I found a kind of problem caused by that: the songs are added in a different order compared to what YTM returns. So, if you sort by date added, it's not identical to what YTM responds.

This I think is fine except for likes, I find it annoying that if I order songs by liked date, it's not identical to YTM. What do you think? Should we revert the corutine at the end of syncLikedSongs() and just leave a for each for adding the likes? 
![image](https://github.com/user-attachments/assets/69c4e626-5bc3-49e8-ba80-b8cf029e4d99)

Another option I see is to do first the toggle like logic linearly and then insert/update with a corutine

mattcarter11 committed:
app: add a global LocalNetworkStatus

mattcarter11 committed:
artis: improve remote/local screen

- If lost connection, change to local screen
- If no connection, don't change to remote
- fix local header being ellipsed when it shouldn't

mikooomich commented:
@mattcarter11 

> Another option I see is to do first the toggle like logic linearly and then insert/update with a corutine


IIRC toggle like just assigns a timestamp, so that would be a best of both worlds solution. Though, I'd have some thread limit (ex. thread pools) for jobs like this, if dispatchers doesn't do this automatically. 

But also...

Does coroutines can provides significant performance boost over just a simple for? The db transactions are processed sequentially anyways. I'd imagine you put the `val localSong = ` outside of the transaction because it deadlock(?).  Maybe putting everything in one transaction and doing everything in a regular for loop may yield better results, and avoid such deadlocks?


I haven't had the time to look at this too closely, but in general, do what you think would be best; you probably are more familiar with this code than me. Thanks for doing the sync work I dread lol

mattcarter11 commented:
@mikooomich okey. For now, I'll let the likes sync as a simple for loop. Later we can improve this. If the rest of the changes seem ok could you merge the PR? I have other changes on a branch based on this one but related to other stuff that i would like to open a PR.

Just for spoliers.....🤫🤫
<img src="https://github.com/user-attachments/assets/09dee3e5-9614-4ec0-b3c1-f20e39ff6b9b" width="200"/>

mikooomich committed:
strings: Move sort_by_line_date to OuterTune section

mikooomich commented:
Yeah, I'll merge it and later on make stuff efficient and/or fix bugs.

Remote artists have a little ui issue when offline. Maybe squeeze this fix in somewhere, or I'll get around to it... eventually
![image](https://github.com/user-attachments/assets/4edc0c7f-57af-4dd5-8e98-e3f8797e5580)

mikooomich commented on app/src/main/java/com/dd3boh/outertune/ui/screens/artist/ArtistScreen.kt:
```diff
@@ -125,176 +131,274 @@ fun ArtistScreen(
+        if (!showLocal) {
+            showLocal = true
+        }
+    }
```
@mattcarter11 do you know why this is done? Doesn't this always set showLocal to true? Currently, when online, only local content will be shown, and to my knowledge, the online content is never shown.


From what I gather, you want to show offline content when offline and online content when online?

mikooomich replied:
Also, I would like to merge in the compose 1.7 changes soon (since they'll probably end up conflicting badly), are you ok with me merging that in after your 3 prs (https://github.com/DD3Boh/OuterTune/pull/127 https://github.com/DD3Boh/OuterTune/pull/125 https://github.com/DD3Boh/OuterTune/pull/121)? 

Here is the branch with compose changes rebased on current dev https://github.com/mikooomich/OuterTune/tree/canary

mattcarter11 replied:
@mikooomich ow, it's missing some logic, here's a fix: https://github.com/DD3Boh/OuterTune/pull/128
About the merge, yes, currently I don't have new code to PR (but I will) so all good for me
